### DISCUSSION TITLE
ListItemAllFields string converted to a date

### DISCUSSION THREAD

**Discussion by sabrinadeveloper2023** [pnp/pnpcore#1571] [General] [answered]
Hi,

I have a custom field named “Number”, the type of the field is string, with the following text: "0004603-4-1"

When I retrieve the field in the collection ListItemAllFields, I see the library is converting the value to a Date.

How can I retrieve the raw values using the same collection?

Please see attachemnt with the [example/image]([doc1.docx](https://github.com/user-attachments/files/17683627/doc1.docx))

Thanks

**Comment by sabrinadeveloper2023** [comment]
[doc1.docx](https://github.com/user-attachments/files/17683633/doc1.docx)

**Comment by sabrinadeveloper2023** [comment]
Hi,

Can someone please help me with this issue? This is causing issues in production

Thanks

**Comment by jansenbe** [comment]
@sabrinadeveloper2023 : if the string can be parsed as DateTime it's returned as such (see https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Services/Core/JsonMappingHelper.cs#L1049-L1058). The code path that does this has no knowledge about the FieldType at the moment of parsing, hence this best effort approach. Structural approach would be to ensure it has access to the field types, for now can you see if loading and using the `FieldValuesAsText` property helps. 

Below snippet was copied from one of our test cases:

```csharp
var myList2 = context2.Web.Lists.FirstOrDefault(p => p.Title == listTitle);
await myList2.LoadAsync(p => p.Items.QueryProperties(p => p.Title, p => p.FieldValuesAsText));

foreach (var item2 in myList2.Items.AsRequested())
{
    Assert.IsTrue(item2.Title == "okido");
    Assert.IsTrue(item2.FieldValuesAsText.Values["MetaInfo"].ToString().Contains("{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|MyStringWithXmlValues"));
}
```

**Comment by sabrinadeveloper2023** [comment]
@jansenbe Thanks for your answer. I have implemented your suggestions and the custom field I need to retrieve is not part of the collection item.FieldValuesAsText.Values.

Please let me know what do you think as alternative solution to retrieve the value in the custom field as text.

Thanks for your help, much appreciated

**Reply by Magnus-K-work** [reply, marked as answer]
As I'm learning PnP.Core from scratch, I tried to replicate this issue and I can see @sabrinadeveloper2023's issue as well.

Steps to replicate the issue:

1. Add a blank list in a site collection (I called mine "CustomList")
2. In the new list, add a column that takes a single-row string (mine was called "CustomCol")
3. Add a list item, set title to "ItemA" and value in the CustomCol field to "0004603-4-1" (the value in sabrinadeveloper2023's example above)
4. In the code, fetch the list and iterate through the list items.
5. For each list item, load the FieldValuesAsText property
6. Compare FieldValuesAsText and Values properties
7. FieldValuesAsText does not contain any of the columns that Values contains? 

My code:
```
using (var context = await _pnpContextFactory.CreateAsync(siteUri))
{
    var customList = await context.Web.Lists.GetByTitleAsync("CustomList");
    foreach (var listItem in customList.Items)
    {
        await listItem.LoadAsync(p => p.FieldValuesAsText);
        var fieldValuesAsText = listItem.FieldValuesAsText;
        var allValues = listItem.Values;
    }
}
```
Comparison of FieldValuesAsHtml and Values properties as a screenshot. Note the value of the CustomCol column to the right, the expected raw string value is "0004603-4-1":
![screenshot](https://github.com/user-attachments/assets/71b72859-2351-49b3-a9ed-1530a86a2ba3)

**Reply by Magnus-K-work** [reply]
@jansenbe I don't think my reply is an answer, but more a detailed step-by-step recreation of the problem. It ends with a problem where ListItem.Values is a completely different set of key-values than ListItem.FieldValuesAsText.

**Comment by sabrinadeveloper2023** [comment]
@Magnus-K-work Thanks for reproducing this problem! Much appreciated your input as this is causing a big issue in production and I am not sure what is the best approach ☹. @jansenbe I have tried several approaches but none of them works. Thanks.

**Reply by Magnus-K-work** [reply]
I tried some more ideas:
- Failed with FieldValuesAsHtml which contains same fields as FieldValuesAsText which doesn't contain the list items's fields
- Failed with FieldValuesForEdit, which seems to be empty even if I tried to load that property.
👎 

My suggested solution:
So, I'd try to fetch the values using Microsoft Graph API instead but I do understand if this will be a bunch of extra code as you need to get a Graph client first. I used auth provider for a certificate:
```
private GraphServiceClient GetGraphServiceClient()
{
    var certificate = new X509Certificate2(<Path to PFX certificate file>, <certificate password>);
    var authenticationProvider = new ClientCertificateCredential(
        clientId: <Client ID>,
        tenantId: <Tenant ID>,
        clientCertificate: certificate);

    var scopes = new[] { "https://graph.microsoft.com/.default" };

    var graphServiceClient = new GraphServiceClient(authenticationProvider, scopes);

    return new GraphServiceClient(authenticationProvider);
}
```
Then all the stuff to get down to a ListItem from Graph, remember to set the correct API permissions for Graph for this app registration:
```
// Fetch the site
var theSite = await graphClient.Sites["<site>.sharepoint.com:/sites/<site name>"].GetAsync();

// Fetch the list named "CustomList" from the site
var customList = await graphClient.Sites[theSite.Id].Lists["CustomList"].GetAsync();

// Fetch the items in the list
var listItems = await graphClient.Sites[theSite.Id].Lists[customList.Id].Items.GetAsync((requestConfiguration) =>
{
    requestConfiguration.QueryParameters.Expand = new string[] { "fields" };
});

foreach (var item in listItems.Value)
{
    // Access item properties
    var itemId = item.Id;
    var itemFields = item.Fields.AdditionalData;

    foreach (var field in itemFields)
    {
        Console.WriteLine($"{field.Key}: {field.Value}");
    }
}
```
The items.Fields.AdditionalData seems to contain the correct values for CustomCol:
![graph_field_values](https://github.com/user-attachments/assets/51d363e9-6b41-407f-871d-c3a996425e77)

HTH!
Magnus
